File: JAVA/1140. Stone Game II.java
class Solution {
    public int solve(int [] piles,int i,int trn,int m,int [][][]dp)
    {
        int n=piles.length;
        if(i>=n)
            return 0;
        if(dp[trn][i][m]!=0)
            return dp[trn][i][m];
        int res=trn==1?Integer.MAX_VALUE:Integer.MIN_VALUE;
        int scr=0;
        for(int x=1;x<=Math.min(2*m,n-i);x++)
        {
            scr+=piles[i+x-1];
            if(trn==0)
                res=Math.max(res,scr+solve(piles,i+x,1,Math.max(m,x),dp));
            else                 
                res=Math.min(res,solve(piles,i+x,0,Math.max(m,x),dp));        
        }
        return dp[trn][i][m]=res;
    }
    public int stoneGameII(int[] piles) {
        int [][][]dp=new int[2][piles.length+1][piles.length+1];
        return solve(piles,0,0,1,dp);
    }
}

File: C++/1140. Stone Game II.cpp
class Solution {
public:
    int stoneGameII(vector<int>& piles) {
        int length = piles.size();
        vector<vector<int>> dp(length + 1, vector<int>(length + 1,0));
        vector<int> sufsum (length + 1, 0);
        for (int i = length - 1; i >= 0; i--) {
            sufsum[i] = sufsum[i + 1] + piles[i];
        }
        return helper(dp, sufsum, 0, 1);
    }    
    int helper(vector<vector<int>>& dp, vector<int>& sufsum, int i, int M) {
        if (i == sufsum.size()) return 0;
        if (2*M >= sufsum.size() - i) return sufsum[i];
        if (dp[i][M] != 0) return dp[i][M];
        int res = INT_MAX;
        for (int X = 1; X <= 2*M; X++) {
            res = min(res, helper(dp, sufsum, i+X, max(X,M)));
        }
        dp[i][M] = sufsum[i] - res;
        return dp[i][M];
    }
};
